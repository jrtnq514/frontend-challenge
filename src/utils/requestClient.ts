import ky from "ky";

const API_BASE_URL = process.env.API_BASE_URL || "http://localhost:3001/api";

export default ky.create({
  headers: {},
  prefixUrl: API_BASE_URL,
});
