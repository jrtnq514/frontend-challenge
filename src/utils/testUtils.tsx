import { InitialEntry, RouterState } from "@remix-run/router";
import {
  createMemoryRouter,
  RouteObject,
  RouterProvider,
} from "react-router-dom";
import {
  SignUpState,
  SignUpStateProvider,
} from "src/contexts/signUpStateContext";
import { render } from "@testing-library/react";

interface RenderViewProps {
  routes: RouteObject[];
  routerOptions?:
    | {
        basename?: string | undefined;
        hydrationData?:
          | Partial<Pick<RouterState, "errors" | "loaderData" | "actionData">>
          | undefined;
        initialEntries?: InitialEntry[] | undefined;
        initialIndex?: number | undefined;
      }
    | undefined;
  signUpState?: SignUpState;
}

export const renderView = ({
  routes,
  routerOptions = {},
  signUpState,
}: RenderViewProps) => {
  const router = createMemoryRouter(routes, routerOptions);

  return {
    ...render(
      <SignUpStateProvider initState={signUpState}>
        <RouterProvider router={router} />
      </SignUpStateProvider>
    ),
    router,
  };
};
