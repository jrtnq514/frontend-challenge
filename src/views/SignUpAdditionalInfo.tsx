import React, { useEffect } from "react";
import { Link, useLoaderData, useNavigate } from "react-router-dom";
import {
  Box,
  Heading,
  VStack,
  HStack,
  Button,
  Text,
  FormControl,
  FormLabel,
  FormErrorMessage,
  Select,
  Checkbox,
} from "@chakra-ui/react";
import { useSignUpState } from "src/contexts/signUpStateContext";
import { useForm, SubmitHandler } from "react-hook-form";

export interface SignUpAdditionalInfoInputs {
  favoriteColor: string;
  agreeTermsAndConditions: boolean;
}

const SignUpAdditionalInfo: React.FC = () => {
  const navigate = useNavigate();
  const { signUpState, setSignUpState } = useSignUpState();
  const {
    register,
    handleSubmit,
    setValue,
    formState: { errors },
  } = useForm<SignUpAdditionalInfoInputs>({ defaultValues: { agreeTermsAndConditions: false }});
  const colors = useLoaderData() as string[];

  useEffect(() => {
    if (!signUpState.firstName || !signUpState.email || !signUpState.password) {
      navigate("/");
    }

    if (signUpState.favoriteColor) {
      setValue("favoriteColor", signUpState.favoriteColor);
    }
  }, []);

  const onSubmit: SubmitHandler<SignUpAdditionalInfoInputs> = (data) => {
    setSignUpState({
      ...signUpState,
      favoriteColor: data.favoriteColor,
      agreeTermsAndConditions: data.agreeTermsAndConditions
    });
    navigate("/confirmation");
  };

  return (
    <Box>
      <VStack spacing={6}>
        <Heading as="h1">Sign Up</Heading>
        <Text>Lets get started with some basic information.</Text>
        <Box as="form" width="100%" onSubmit={handleSubmit(onSubmit)}>
          <VStack spacing={4}>
            <FormControl isInvalid={!!errors.favoriteColor}>
              <FormLabel>Choose your favorite color</FormLabel>
              <Select
                placeholder="Choose a color"
                {...register("favoriteColor", { required: true })}
              >
                { colors.map((color) => <option value={color} key={color}>{color}</option>) }
              </Select>
              {errors.favoriteColor && errors.favoriteColor.type === "required" && (
                <FormErrorMessage>Favorite color is required.</FormErrorMessage>
              )}
            </FormControl>
            <FormControl isInvalid={!!errors.agreeTermsAndConditions}>
              <HStack>
                <Checkbox {...register("agreeTermsAndConditions", { required: true })}>
                  I Agree to <Link to="#">Terms and Conditions</Link>
                </Checkbox>
              </HStack>
              <FormErrorMessage>You must agree to the terms and conditions.</FormErrorMessage>
            </FormControl>
            <HStack justifyContent="space-between" width="100%">
              <Button type="button" variant="outline" colorScheme="green" onClick={(event) => navigate('/')}>Back</Button>
              <Button type="submit" variant="solid" colorScheme="green">
                Next
              </Button>
            </HStack>
          </VStack>
        </Box>
      </VStack>
    </Box>
  );
};

export default SignUpAdditionalInfo;
