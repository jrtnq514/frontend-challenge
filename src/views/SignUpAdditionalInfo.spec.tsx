import SignUpAdditionalInfo from "./SignUpAdditionalInfo";
import { waitFor, screen } from "@testing-library/react";
import { renderView } from "src/utils/testUtils";

describe("SignUpAdditionalInfo", () => {
  const mockColors = ["Blue", "Black", "Red"];

  it("should render dropdown with correct colors", async () => {
    const { container } = renderView({
      routes: [
        {
          path: "/more-info",
          element: <SignUpAdditionalInfo />,
          loader: () => mockColors,
        },
      ],
      routerOptions: { initialEntries: ["/more-info"] },
      signUpState: { firstName: "test", email: "test", password: "test" },
    });

    await waitFor(() => screen.findAllByTestId("color-select-option"));

    const colorOptions = (await screen.findAllByTestId(
      "color-select-option"
    )) as HTMLOptionElement[];

    expect(colorOptions.map((x) => x.value)).toEqual(mockColors);
    expect(container).toMatchSnapshot();
  });

  it("should redirect if certain states are not set", async () => {
    renderView({
      routes: [
        {
          path: "/more-info",
          element: <SignUpAdditionalInfo />,
          loader: () => mockColors,
        },
        {
          path: "/",
          element: <div>Redirected</div>,
        },
      ],
      routerOptions: { initialEntries: ["/more-info"] },
    });

    await waitFor(() => {
      screen.findByText("Redirected");
    });

    expect(await screen.findByText("Redirected")).not.toBeNull();
  });
});
