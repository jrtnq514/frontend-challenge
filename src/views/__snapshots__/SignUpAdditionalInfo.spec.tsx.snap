// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpAdditionalInfo should render dropdown with correct colors 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="chakra-stack css-pfcv3f"
    >
      <h1
        class="chakra-heading css-0"
      >
        Additional Information
      </h1>
      <p
        class="chakra-text css-0"
      >
        Just a few more things.
      </p>
      <form
        class="css-8atqhb"
        data-testid="more-info-form"
      >
        <div
          class="chakra-stack css-pfcv3f"
        >
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="field-:r0:"
              id="field-:r0:-label"
            >
              Choose your favorite color
            </label>
            <div
              class="chakra-select__wrapper css-42b2qy"
            >
              <select
                class="chakra-select css-1ik61og"
                data-testid="color-select"
                id="field-:r0:"
                name="favoriteColor"
              >
                <option
                  value=""
                >
                  Choose a color
                </option>
                <option
                  data-testid="color-select-option"
                  value="Blue"
                >
                  Blue
                </option>
                <option
                  data-testid="color-select-option"
                  value="Black"
                >
                  Black
                </option>
                <option
                  data-testid="color-select-option"
                  value="Red"
                >
                  Red
                </option>
              </select>
              <div
                class="chakra-select__icon-wrapper css-162mkon"
              >
                <svg
                  aria-hidden="true"
                  class="chakra-select__icon"
                  focusable="false"
                  role="presentation"
                  style="width: 1em; height: 1em; color: currentColor;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <div
              class="chakra-stack css-84zodg"
            >
              <label
                class="chakra-checkbox css-1elxtiq"
                data-testid="terms-checkbox"
              >
                <input
                  aria-disabled="false"
                  aria-invalid="false"
                  class="chakra-checkbox__input"
                  id="field-:r1:"
                  name="agreeTermsAndConditions"
                  style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                  type="checkbox"
                  value=""
                />
                <span
                  aria-hidden="true"
                  class="chakra-checkbox__control css-dnty2r"
                />
                <span
                  class="chakra-checkbox__label css-1oeb2oe"
                >
                  I Agree to
                   
                  <a
                    class="chakra-link css-bd5f8p"
                    href="https://google.com"
                    rel="noopener"
                    target="_blank"
                  >
                    Terms and Conditions
                  </a>
                </span>
              </label>
            </div>
          </div>
          <div
            class="chakra-stack css-1cqqob7"
          >
            <button
              class="chakra-button css-4xx2wk"
              type="button"
            >
              Back
            </button>
            <button
              class="chakra-button css-4xx2wk"
              type="submit"
            >
              Next
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
